### PR TITLE
8266528: Optimize C2 VerifyIterativeGVN execution time

### PR DESCRIPTION
**Pull request by huishi-hs**
Please help review this enhancement for VerifyIterativeGVN, reduce about 3x - 200x executime time when VerifyIterativeGVN is on.

In simple test "-Xcomp -XX:+VerifyIterativeGVN -XX:-TieredCompilation -version", time reduced from 8.67s to 2.4s.
In extreme case hotspot/test/jtreg/compiler/escapeAnalysis/Test6689060.java, time reduced from 20000s to 95s.

Test with "-Xbatch -XX:+VerifyIterativeGVN -XX:-TieredCompilation", tier1/2/3 with fastdebug and no regression.

1. Remove node_arena()->contains checking for verifing  nodes.  _verify_window  is reset before every PhaseIterGVN::optimize. Searching from root or nodes in _verify_window  will not meet nodes whose _idx is not  unique (PhaseIterGVN::optimize is not triggered in the middle of PhaseRenumberLive ).  Assertion every node is in current node_arena() in Node::verify, passes tier1/2/3 checks (with -Xbatch -XX:+VerifyIterativeGVN -XX:-TieredCompilation), no assertion failure happens.

2. Combine verification for nodes in _verify_window into one worklist and skipping redundant nodes in _verify_window.

3. Optimize duplicate checking for same input nodes, skipping if current input index is not its first occurence.

4. Optimize field access: Replace "n->in(j)" with "n->_in[j]", same with outcnt calucation for input node x.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266528](https://bugs.openjdk.java.net/browse/JDK-8266528): Optimize C2 VerifyIterativeGVN execution time

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 9635c9b98a5ec179a366180506e2619d4e48a926
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4045/head:pull/4045` \
`$ git checkout pull/4045`

Update a local copy of the PR: \
`$ git checkout pull/4045` \
`$ git pull https://git.openjdk.java.net/jdk pull/4045/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4045`

View PR using the GUI difftool: \
`$ git pr show -t 4045`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4045.diff">https://git.openjdk.java.net/jdk/pull/4045.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/node.cpp [hunk 1]
```diff
@@ -2236,17 +2230,27 @@ void Node::verify(Node* n, int verify_depth) {
...
-      for (uint j = 0; j < n->len(); j++) {
-        if (n->in(j) == x) {
+      int cnt = 1;
+      for (uint j = 0; j < i; j++) {
+        if (n->_in[j] == x) {
+          cnt++;
+          break;
+        }
+      }
+      if (cnt == 2) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/node.cpp:2249]: I think it should be `cnt > 1`.

File: test/hotspot/jtreg/compiler/debug/TraceIterativeGVN.java [hunk 2]
```diff
@@ -23,7 +24,7 @@
 
 /*
  * @test
- *
+ * @requires vm.debug == true & vm.flavor == "server"
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/debug/TraceIterativeGVN.java:27]: Use `vm.compiler2.enabled` instead of `vm.flavor == "server"`

**Review by vnkozlov — CHANGES_REQUESTED** [review]

File: src/hotspot/share/opto/node.cpp (hunk 1 above)
Comment (huishi-hs) [reply src/hotspot/share/opto/node.cpp:2249]: previous loop breaks when meet first input which is same with x, so cnt must be 2 if x is duplicated with previous input.

File: test/hotspot/jtreg/compiler/debug/TraceIterativeGVN.java (hunk 2 above)
Comment (huishi-hs) [reply test/hotspot/jtreg/compiler/debug/TraceIterativeGVN.java]: Got! updated in new commit.

File: src/hotspot/share/opto/node.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/node.cpp:2249]: You are right. I missed `break`. Using an other local VectorSet here may not be cheeper. Okay.
Add comment here. Something about that we processed this node already.

**Review by vnkozlov — APPROVED** [review]

File: src/hotspot/share/opto/node.cpp (hunk 1 above)
Comment (huishi-hs) [reply src/hotspot/share/opto/node.cpp:2249]: Thanks for your review! comment is added.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Comment by huishi-hs** [comment]
@TobiHartmann Thanks!

**Pull request closed without merging**